diff --git a/packages/apps/Settings/src/com/android/settings/gestures/GesturesSettingPreferenceController.java b/packages/apps/Settings/src/com/android/settings/gestures/GesturesSettingPreferenceController.java
index 8c2d64b..e97a47b 100644
--- a/packages/apps/Settings/src/com/android/settings/gestures/GesturesSettingPreferenceController.java
+++ b/packages/apps/Settings/src/com/android/settings/gestures/GesturesSettingPreferenceController.java
@@ -80,6 +80,7 @@ public class GesturesSettingPreferenceController extends BasePreferenceControlle
         controllers.add(new SwipeToScreenshotPreferenceController(context, FAKE_PREF_KEY));
         controllers.add(new ScreenOffUdfpsPreferenceController(context, FAKE_PREF_KEY)
                 .setConfig(ambientDisplayConfiguration));
+        controllers.add(new VolumeButtonMusicControlPreferenceController(context, FAKE_PREF_KEY));
         return controllers;
     }
 }
