diff --git a/build/make/target/product/handheld_product.mk
index 2199c57..e89c702 100644
--- a/build/make/target/product/handheld_product.mk
+++ b/build/make/target/product/handheld_product.mk
@@ -34,7 +34,8 @@ PRODUCT_PACKAGES += \
     preinstalled-packages-platform-handheld-product.xml \
     QuickSearchBox \
     SettingsIntelligence \
-    frameworks-base-overlays
+    frameworks-base-overlays \
+    Via
 
 PRODUCT_PACKAGES_DEBUG += \
     frameworks-base-overlays-debug
