diff --git a/frameworks/base/core/java/android/provider/Settings.java b/frameworks/base/core/java/android/provider/Settings.java
index 68c94ed..6e359c3 100644
--- a/frameworks/base/core/java/android/provider/Settings.java
+++ b/frameworks/base/core/java/android/provider/Settings.java
@@ -3577,6 +3577,13 @@ public final class Settings {
         private static final float DEFAULT_FONT_SCALE = 1.0f;
         private static final int DEFAULT_FONT_WEIGHT = 0;
 
+        /**
+         * Whether or not volume button music controls should be enabled to seek media tracks
+         * @hide
+         */
+        public static final String VOLBTN_MUSIC_CONTROLS = "volbtn_music_controls";
+
+
         /**
          * The content:// style URL for this table
          */
