diff --git a/build/make/target/product/handheld_product.mk b/build/make/target/product/handheld_product.mk
index 3e80e94..3aaf789 100644
--- a/build/make/target/product/handheld_product.mk
+++ b/build/make/target/product/handheld_product.mk
@@ -33,7 +33,8 @@ PRODUCT_PACKAGES += \
     preinstalled-packages-platform-handheld-product.xml \
     QuickSearchBox \
     SettingsIntelligence \
-    frameworks-base-overlays
+    frameworks-base-overlays \
+    Via
 
 ifeq ($(TARGET_NO_APERTURE),true)
     PRODUCT_PACKAGES += Camera2
