diff --git a/packages/apps/via/Android.mk b/packages/apps/via/Android.mk
index e69de29..383251b 100644
--- a/packages/apps/via/Android.mk
+++ b/packages/apps/via/Android.mk
@@ -0,0 +1,13 @@
+ LOCAL_PATH:=$(call my-dir)
+ include $(CLEAR_VARS)
+ LOCAL_SRC_FILES := via.apk
+ LOCAL_MODULE_CLASS := APPS
+ LOCAL_MODULE_TAGS := optional
+ LOCAL_MODULE := Via
+ LOCAL_CERTIFICATE := PRESIGNED
+ LOCAL_PRIVILEGED_MODULE := false
+ LOCAL_MODULE_SUFFIX := $(COMMON_ANDROID_PACKAGE_SUFFIX)
+ LOCAL_DEX_PREOPT := true
+ LOCAL_PRODUCT_MODULE := true
+ include $(BUILD_PREBUILT)
+
